### ISSUE TITLE
Assert macro causes StackOverflowError on recursive (?) data structures

### ISSUE THREAD

**Issue by hugo-vrijswijk** [scalatest/scalatest#1871]
Scastie reproduction: https://scastie.scala-lang.org/TltttkpOQC22q3fNvV7n6A

Comparing two recursive data structures (like an AST from scalameta) with `org.scalatest.Assertions#assert` produces a StackOverflowError. I _think_ this is an issue with ScalaTest as I don't expect any StackOverflowError when comparing two values.

Expected: a `TestFailedException` with a comparison between the two values

Reproduction code:

```scala
import scala.meta._

val first = q"a(1)"
val second = q"b(2)"

try {
  org.scalatest.Assertions.assert(first == second)
} catch {
  case _: StackOverflowError => println("Stackoverflow!!!")
}

first == second
println("No stackoverflow!!!")
```

**Comment by cheeseng** [comment]
@hugo-vrijswijk  I just released a 3.2.2-M2, do you mind to try to see if it fixes your problem?

Cheers.

**Comment by hugo-vrijswijk** [comment]
As the StackOverflow has been fixed I'll close this issue and create another for the large output so the discussion can be focused there. Thanks a lot for the quick investigation and fix!
